feat: allow to import local models

lstocchi opened the pull request:
### What does this PR do?

It allows to import .gguf models that have already been downloaded by the user and that are on his machine.
After imported, the user can start a service, open its folder and delete it

### Screenshot / video of UI

![import_model](https://github.com/containers/podman-desktop-extension-ai-lab/assets/49404737/689c15b2-94f7-4529-b2f5-982b663e9dfe)


### What issues does this PR fix or reference?

it resolves #173 

Fixes #814 


### How to test this PR?

1. download a gguf model which is not listed in the catalog (e.g. take the first ones from here https://huggingface.co/TheBloke/Mistral-7B-Instruct-v0.1-GGUF/tree/main 
2. go to models, import, select the gguf file
3. once imported, try to play with it (inference server + playground)
4. then delete it (it will be removed from catalog and disk)

slemeur commented:
Nice!

lstocchi commented:
There is an issue when you delete a model having an inference server using it and you stop/start ai lab again. Fixing it

axel7083 commented on packages/shared/src/uri/Uri.ts:
```diff
@@ -0,0 +1,78 @@
+ ***********************************************************************/
+import type { Uri as APIUri } from '@podman-desktop/api';
+
+export class Uri {
```
Could you explain why we need to have our own implementation of the Uri class ? I am not sure to get it

lstocchi replied:
The Uri object which is returned when calling the api is serialized using private properties bc the Uri constructor uses `private _scheme ..., private _path ...`. So when we receive the resulting object in ai lab we have 
```
_path: '...',
_scheme: '...',
```
In this way it is not kinda usable so we need to "revive" it so that its fields are exposed through methods like the original Uri.
The same thing is done on Desktop.

axel7083 replied:
What are the properties that we need ?

axel7083 commented on packages/backend/src/managers/inference/inferenceManager.ts:
```diff
@@ -383,11 +392,12 @@ export class InferenceManager extends Publisher<InferenceServer[]> implements Di
             },
             status: containerInfo.Status === 'running' ? 'running' : 'stopped',
-            models: modelsId.map(id => this.modelsManager.getModelInfo(id)),
+            models: models,
```
I think we can avoid having all this try catch with the `getModelInfo` potentially throwing error by filtering the models Id using the `isModelOnDisk` from the `ModelsManager`

```suggestion
            models: modelsId.filter(id => this.modelsManager.isModelOnDisk(id)).map(id => this.modelsManager.getModelInfo(id)),,
```

lstocchi replied:
Your suggestion does a different thing but i don't know which should be the correct behavior tbh.

In the current code i skip the containers that are associated to a model which is not on the catalog/disk anymore.
In your case you're still detecting the inference server but its models list can be empty.

So not sure how it should work. Should we not show the inference servers that have been created with a no-more-existing model or we should show them and associate an empty models list? Not sure if the latter implies some misbehavior though.

lstocchi replied:
Agree, opened https://github.com/containers/podman-desktop-extension-ai-lab/issues/791

axel7083 replied:
SO for now I would rather have the servers showed even tho we don't have the model, so the user can do cleanup from ai studio

lstocchi replied:
Done

axel7083 commented on packages/backend/src/managers/catalogManager.ts:
```diff
@@ -93,4 +104,38 @@ export class CatalogManager extends Publisher<ApplicationCatalog> implements Dis
+      });
+    }
+
+    const customCatalog = path.resolve(this.appUserDirectory, 'catalog.json');
```
I think we should have two catalog here. I am not in favor of overwriting the _native_ catalog, I feel it would be more appropriate to have some kind of `user-models.json` in the `appUserDirectory` what do you think ?

lstocchi replied:
From my point of view the operation that the user is doing here it is just updating the custom catalog.json file that we already support. It is just that we guide them. But if someone wants to add a new model manually we are already saying that they need to create a `catalog.json` file (copy/paste the `assets/ai.json`) inside the ai-lab folder and update it. 

So i wouldn't create a new file as it may overcomplicate stuff, but let's see what others think

axel7083 replied:
How do you handle the case where we update on our side the catalog ? Are we losing the user imported models ?

lstocchi replied:
When importing, if the `catalog.json` file already exists, it just adds the new models in there and update the file on disk. If it does not exists it creates it on disk by copying the content from the native `ai.json` and adding the imported models. 

So if the user wants to update the catalog after having imported models he would work as before. Open the `catalog.json` file which would also contain the imported models and edit it.

axel7083 replied:
What if some day we want to remove a model ? like we did with llama2 ? the mechanism you describe would not be suitable ?

lstocchi replied:
This is the same thing that may happen today. If i create my custom catalog.json and we update ai.json, ai-lab will still point to the custom catalog. We have to refactor how the custom catalog + ai.json work or inform the user any update. Opening an issue

lstocchi replied:
-> https://github.com/containers/podman-desktop-extension-ai-lab/issues/792

jeffmaury replied:
What about having the 2 files but merging them instead of choosing one over the other

axel7083 reviewed (commented):
Nice feature ! I added some comment

lstocchi commented on packages/backend/src/managers/catalogManager.ts:
```diff
@@ -59,9 +65,14 @@ export class CatalogManager extends Publisher<ApplicationCatalog> implements Dis
 
+  onCatalogUpdate(listener: catalogUpdateHandle): void {
+    this.#catalogUpdateListeners.push(listener);
+  }
```
Done, Thanks!!

axel7083 reviewed (commented):
Almost looking good just some stuff I find out while testing it

## Breadcrumb

<img width="787" alt="image" src="https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/95e06ec3-bad2-4fd2-a505-ec333042bb4e">

Could we have the breadcrumb allowing to go back to models page ?

## Import timeout

Taking very long time to select a file in explorer gave a timeout

<img width="901" alt="image" src="https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/e0776bcb-4046-4152-bcee-e0f99c53c184">

We probably need to find out some kind of way to get back this value and update the form. (This could be skipped, but would need to be address at some point)

## Selected file

Maybe I should not be able to select files inside the models folder ?

<img width="779" alt="image" src="https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/986186f6-78a1-4e74-8f6f-9721ab358da6">

If I import my own file, it gets weird

<img width="575" alt="image" src="https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/2ef88e56-dc69-491c-9a91-9ea666190ead">

lstocchi commented:
Nice catches!!

> Almost looking good just some stuff I find out while testing it
> 
> ## Breadcrumb
 
Done

> ## Import timeout
> Taking very long time to select a file in explorer gave a timeout

I'll give a look on desktop side as the timeout is really tight. For the moment i added an error message if something unexpected happen so the user is informed.

> ## Selected file
> Maybe I should not be able to select files inside the models folder ?

Added a check to verify that the selected model is not already in the catalog.

jeffmaury commented:
The size information is missing and I'm wondering how the RAM usage is computed

jeffmaury reviewed (commented):
The timeout issue is to be fixed because it breaks the user experience

lstocchi commented:
> The size information is missing and I'm wondering how the RAM usage is computed

The RAM usage is = size of the file as it seems to me that that is the value we use in the catalog, no?

lstocchi commented:
@jeffmaury timeout removed and size + creation date added

axel7083 commented on packages/shared/src/messages/MessageProxy.ts:
```diff
@@ -124,6 +124,7 @@ export class RpcBrowser {
   counter: number = 0;
   promises: Map<number, { resolve: (value: unknown) => unknown; reject: (value: unknown) => void }> = new Map();
   subscribers: Map<string, (msg: any) => void> = new Map();
+  noTimeoutChannels: string[] = ['openDialog'];
```
please make it available somewhere like the Messages ids in consts file or something

lstocchi replied:
Done

feloy commented:
> > The size information is missing and I'm wondering how the RAM usage is computed
> 
> The RAM usage is = size of the file as it seems to me that that is the value we use in the catalog, no?

No, the ram usage does not come from the file size. For the models in the catalog, it is computed with the script in `.github/workflows/compute-model-sizes.yaml` (see #674)

lstocchi commented:
> > > The size information is missing and I'm wondering how the RAM usage is computed
> > 
> > 
> > The RAM usage is = size of the file as it seems to me that that is the value we use in the catalog, no?
> 
> No, the ram usage does not come from the file size. For the models in the catalog, it is computed with the script in `.github/workflows/compute-model-sizes.yaml` (see #674)

Ah I completely missed that part. Thanks for telling me. Removed the memory property for the moment then. 👍

axel7083 commented on packages/shared/src/models/NoTimeoutChannels.ts:
```diff
@@ -0,0 +1,19 @@
+/**********************************************************************
```
Why is it in the `src/models` folder ? could we move it to the messages folder ?

axel7083 commented on packages/frontend/src/pages/ImportModels.svelte:
```diff
@@ -0,0 +1,147 @@
+<NavPage lastPage="{{ name: 'Models', path: '/models' }}" title="Import Models" searchEnabled="{false}">
+  <svelte:fragment slot="content">
+    <div class="p-5 min-w-full h-fit">
+      <div class="bg-charcoal-600 px-6 py-4 space-y-2 lg:px-8 sm:pb-6 xl:pb-8">
```
```suggestion
      <div class="py-4 space-y-2">
```

**before**
![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/ff9ece71-a713-4fb5-b90e-ed017eac2907)

**after**
![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/d8cdced8-2a1a-4dfb-9636-39aa6c860331)

axel7083 replied:
```suggestion
      <div class="bg-charcoal-800 rounded-lg px-6 py-4 space-y-2">
```

**before**

![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/9d689a6a-bd3d-4a48-b7d4-ddc851932292)

**after**
![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/e8556fce-5c66-40e4-a25a-066ee5ee8938)

axel7083 commented on packages/frontend/src/pages/ImportModels.svelte:
```diff
@@ -0,0 +1,147 @@
+            <ErrorMessage class="py-2 text-sm" error="{importError}" />
+          {/if}
+        </div>
+        <Button
```
```suggestion
        <div class="pt-5 border-zinc-600 border-t-2"/>
        <Button
```

Adding a divider to match the design container import

lstocchi replied:
So the container import page has to be updated. I forgot and i'll create a PR for this on desktop. When working on the load images page i was copying the container import template and Tim commented to remove the divider as it is not used anywhere else https://github.com/containers/podman-desktop/pull/6540#discussion_r1539272375

axel7083 replied:
Okey np

lstocchi replied:
Thanks for making me remind about it 😆 -> https://github.com/containers/podman-desktop/pull/6655

axel7083 commented on packages/frontend/src/pages/ImportModels.svelte:
```diff
@@ -0,0 +1,147 @@
+<NavPage lastPage="{{ name: 'Models', path: '/models' }}" title="Import Models" searchEnabled="{false}">
+  <svelte:fragment slot="content">
+    <div class="p-5 min-w-full h-fit">
+      <div class="bg-charcoal-800 rounded-lg px-6 py-4 space-y-2">
```
```suggestion
      <div class="bg-charcoal-800 rounded-lg px-6 py-6 space-y-2">
              <span class="block mb-2 text-sm font-medium text-gray-400">Models to import:</span>
```

To match the design of the import containers

Here is the example
![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/06fa97a8-53ed-49e7-9fd9-4641e04d7a14)

with the changes you get
![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/529ec43d-1294-468a-ad8d-dd7f78b9cbb3)

axel7083 commented on packages/frontend/src/lib/table/model/ModelColumnName.svelte:
```diff
@@ -8,9 +8,18 @@ function openDetails() {
+  title="{object.name}"
+  on:click="{() => openDetails()}"
+  aria-label="Open Model Details">
+  <div class="text-sm text-gray-300 overflow-hidden text-ellipsis w-full" aria-label="Model Name">
```
```suggestion
  <div class="text-sm text-gray-300 overflow-hidden text-ellipsis w-full text-left" aria-label="Model Name">
```

When the name is too short we have some weird alignements

![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/9f7f8698-121d-4ee1-bca4-583cb1f1d04b)

jeffmaury commented:
> > > > The size information is missing and I'm wondering how the RAM usage is computed
> > > 
> > > 
> > > The RAM usage is = size of the file as it seems to me that that is the value we use in the catalog, no?
> > 
> > 
> > No, the ram usage does not come from the file size. For the models in the catalog, it is computed with the script in `.github/workflows/compute-model-sizes.yaml` (see #674)
> 
> Ah I completely missed that part. Thanks for telling me. Removed the memory property for the moment then. 👍

I would rather default the memory field to the file size

axel7083 commented:
> > > > > The size information is missing and I'm wondering how the RAM usage is computed
> > > > 
> > > > 
> > > > The RAM usage is = size of the file as it seems to me that that is the value we use in the catalog, no?
> > > 
> > > 
> > > No, the ram usage does not come from the file size. For the models in the catalog, it is computed with the script in `.github/workflows/compute-model-sizes.yaml` (see #674)
> > 
> > 
> > Ah I completely missed that part. Thanks for telling me. Removed the memory property for the moment then. 👍
> 
> I would rather default the memory field to the file size

I think this is misleading

vrothberg commented:
If we don't have the information about the resource constraints of a model, I wouldn't guess it.  Maybe we can display `NA` (not available) for imported models?

jeffmaury commented:
> > > > > > The size information is missing and I'm wondering how the RAM usage is computed
> > > > > 
> > > > > 
> > > > > The RAM usage is = size of the file as it seems to me that that is the value we use in the catalog, no?
> > > > 
> > > > 
> > > > No, the ram usage does not come from the file size. For the models in the catalog, it is computed with the script in `.github/workflows/compute-model-sizes.yaml` (see #674)
> > > 
> > > 
> > > Ah I completely missed that part. Thanks for telling me. Removed the memory property for the moment then. 👍
> > 
> > 
> > I would rather default the memory field to the file size
> 
> I think this is misleading and does not reflect reality

Both are in the same order of magnitude (file size is a little bit bigger) and as we will check the machine size based on this data, I would rather ask a little bit of additional memory than risking failures

axel7083 reviewed (approved):
LGTM ! 🚀 thanks you for your patience and the fixes, great job!

lstocchi commented:
@jeffmaury set the memory value equals to the file size as you said. Let me know if there is something else to change.
Also rebased.

jeffmaury reviewed (approved):
LGTM
